[jit] Setup operator registration for distributed package

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#31214 [jit] Setup operator registration for distributed package**

Summary:

This set up the basic infrastructure for distributed autograd and rpc to
bind their operators to TorchScript, since the whole distributed package
is builtin behind the `USE_DISTRIBUTED` flag, we separate the
registration and build it only when the flag is on.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D19137160](https://our.internmc.facebook.com/intern/diff/D19137160)

wanchaol committed:
[jit] Setup operator registration for distributed package

Summary:

This set up the basic infrastructure for distributed autograd and rpc to
bind their operators to TorchScript, since the whole distributed package
is builtin behind the `USE_DISTRIBUTED` flag, we separate the
registration and build it only when the flag is on.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kostmo commented:
## CircleCI build failures summary
As of commit f06226f:
* **2/2** failures introduced in this PR



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=f06226f95d297764618d4245a621405773bf97da).

### 2 new failures recognized by patterns
The following build failures don't appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3932772) pytorch_linux_xenial_py3_6_gcc5_4_ge_config_legacy_test (1/2)
**Step:** "Test" ([details](https://dr.pytorch.org/build-details.html?build_id=73623726))
```
Dec 12 23:16:42   File "/opt/conda/lib/python3.6/site-packages/torch/jit/__init__.py", line 2046, in _register_builtin 
Dec 12 23:16:42     _get_builtin_table()[id(fn)] = op 
Dec 12 23:16:42   File "/opt/conda/lib/python3.6/site-packages/torch/jit/__init__.py", line 2040, in _get_builtin_table 
Dec 12 23:16:42     _builtin_table[id(dist.autograd.get_gradients)] = "aten::get_gradients" 
Dec 12 23:16:42 AttributeError: module 'torch.distributed' has no attribute 'autograd' 
Dec 12 23:16:42 Traceback (most recent call last): 
Dec 12 23:16:42   File "test/run_test.py", line 456, in <module> 
Dec 12 23:16:42     main() 
Dec 12 23:16:42   File "test/run_test.py", line 449, in main 
Dec 12 23:16:42     raise RuntimeError(message) 
Dec 12 23:16:42 RuntimeError: test_jit_legacy failed! 
Dec 12 23:16:42 + cleanup 
Dec 12 23:16:42 + retcode=1 
Dec 12 23:16:42 + set +x 
Dec 12 23:16:42 =================== sccache compilation log =================== 
Dec 12 23:16:42 =========== If your build fails, please take a look at the log above for possible reasons =========== 
Dec 12 23:16:42 Compile requests                 0 
Dec 12 23:16:42 Compile requests executed        0 
Dec 12 23:16:42 Cache hits                       0 
Dec 12 23:16:42 Cache misses                     0 
Dec 12 23:16:42 Cache timeouts                   0 
```
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3932773) pytorch_linux_xenial_py3_6_gcc5_4_ge_config_simple_test (2/2)
**Step:** "Test" ([details](https://dr.pytorch.org/build-details.html?build_id=73623823))
```
Dec 12 23:16:57   File "/opt/conda/lib/python3.6/site-packages/torch/jit/__init__.py", line 2046, in _register_builtin 
Dec 12 23:16:57     _get_builtin_table()[id(fn)] = op 
Dec 12 23:16:57   File "/opt/conda/lib/python3.6/site-packages/torch/jit/__init__.py", line 2040, in _get_builtin_table 
Dec 12 23:16:57     _builtin_table[id(dist.autograd.get_gradients)] = "aten::get_gradients" 
Dec 12 23:16:57 AttributeError: module 'torch.distributed' has no attribute 'autograd' 
Dec 12 23:16:57 Traceback (most recent call last): 
Dec 12 23:16:57   File "test/run_test.py", line 456, in <module> 
Dec 12 23:16:57     main() 
Dec 12 23:16:57   File "test/run_test.py", line 449, in main 
Dec 12 23:16:57     raise RuntimeError(message) 
Dec 12 23:16:57 RuntimeError: test_jit_simple failed! 
Dec 12 23:16:57 + cleanup 
Dec 12 23:16:57 + retcode=1 
Dec 12 23:16:57 + set +x 
Dec 12 23:16:57 =================== sccache compilation log =================== 
Dec 12 23:16:57 =========== If your build fails, please take a look at the log above for possible reasons =========== 
Dec 12 23:16:57 Compile requests                 0 
Dec 12 23:16:57 Compile requests executed        0 
Dec 12 23:16:57 Cache hits                       0 
Dec 12 23:16:57 Cache misses                     0 
Dec 12 23:16:57 Cache timeouts                   0 
```


---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

wanchaol committed:
Update on "[jit] Setup operator registration for distributed package"


Summary:

This set up the basic infrastructure for distributed autograd and rpc to
bind their operators to TorchScript, since the whole distributed package
is builtin behind the `USE_DISTRIBUTED` flag, we separate the
registration and build it only when the flag is on.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] Setup operator registration for distributed package"


Summary:

This set up the basic infrastructure for distributed autograd and rpc to
bind their operators to TorchScript, since the whole distributed package
is builtin behind the `USE_DISTRIBUTED` flag, we separate the
registration and build it only when the flag is on.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] Setup operator registration for distributed package"


Summary:

This set up the basic infrastructure for distributed autograd and rpc to
bind their operators to TorchScript, since the whole distributed package
is builtin behind the `USE_DISTRIBUTED` flag, we separate the
registration and build it only when the flag is on.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

zdevito commented on torch/jit/__init__.py:
```diff
@@ -2035,6 +2035,10 @@ def register_all(mod):
     if PY37:
         _builtin_table[id(math.remainder)] = "aten::mathremainder"
 
+    import torch.distributed.autograd as dist_autograd
```
This might be better organized as an if statement that appends to the _builtin_ops table, rather than another special case here. In face, all the manually uses of _builtin_table here really should be manual modifications to _buildin_ops.

zdevito commented on torch/csrc/jit/register_distributed_ops.cpp:
```diff
@@ -0,0 +1,60 @@
+  return v.has_value() ? *v : at::Tensor();
+}
+
+c10::OperatorOptions aliasAnalysisFromSchema() {
```
copypasta for this is getting bad, can we directly use the c10 operator registry for these operators? @smessmer can you point to examples of register ops this way.

wanchaol committed:
Update on "[jit] Setup operator registration for distributed package"


Summary:

This set up the basic infrastructure for distributed autograd and rpc to
bind their operators to TorchScript, since the whole distributed package
is builtin behind the `USE_DISTRIBUTED` flag, we separate the
registration and build it only when the flag is on.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] Setup operator registration for distributed package"


Summary:

This set up the basic infrastructure for distributed autograd and rpc to
bind their operators to TorchScript, since the whole distributed package
is builtin behind the `USE_DISTRIBUTED` flag, we separate the
registration and build it only when the flag is on.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

zhaojuanmao commented:
would you please also add DistAutogradJitTestWithSpawn  to test_dist_autograd_fork.py as well? it could be as a follow up diff, thanks
